libsolidity/ast/ASTJsonImporter.cpp

```diff
@@ -0,0 +1,1003 @@
+	if (_json["nodeType"] == "UserDefinedTypeName")
+		return createUserDefinedTypeName(_json);
+	if (_json["nodeType"] == "FunctionTypeName")
+		return createFunctionTypeName(_json);
+	if (_json["nodeType"] == "Mapping")
+		return createMapping(_json);
+	if (_json["nodeType"] == "ArrayTypeName")
+		return createArrayTypeName(_json);
+//	if (_json["nodeType"] == "InlineAssembly")
+//		return createInlineAssembly(_json);
+	if (_json["nodeType"] == "Block")
+		return createBlock(_json);
+	if (_json["nodeType"] == "PlaceholderStatement")
+		return createPlaceholderStatement(_json);
+	if (_json["nodeType"] == "IfStatement")
+		return createIfStatement(_json);
+	if (_json["nodeType"] == "WhileStatement")
+		return createWhileStatement(_json, false);
+	if (_json["nodeType"] == "DoWhileStatement")
+		return createWhileStatement(_json, true);
+	if (_json["nodeType"] == "ForStatement")
+		return createForStatement(_json);
+	if (_json["nodeType"] == "Continue")
+		return createContinue(_json);
+	if (_json["nodeType"] == "Break")
+		return createBreak(_json);
+	if (_json["nodeType"] == "Return")
+		return createReturn(_json);
+	if (_json["nodeType"] == "Throw")
+		return createThrow(_json);
+	if (_json["nodeType"] == "VariableDeclarationStatement")
+		return createVariableDeclarationStatement(_json);
+	if (_json["nodeType"] == "ExpressionStatement")
+		return createExpressionStatement(_json);
+	if (_json["nodeType"] == "Conditional")
+		return createConditional(_json);
+	if (_json["nodeType"] == "Assignment")
+		return createAssignment(_json);
+	if (_json["nodeType"] == "TupleExpression")
+		return createTupleExpression(_json);
+	if (_json["nodeType"] == "UnaryOperation")
+		return createUnaryOperation(_json);
+	if (_json["nodeType"] == "BinaryOperation")
+		return createBinaryOperation(_json);
+	if (_json["nodeType"] == "FunctionCall")
+		return createFunctionCall(_json);
+	if (_json["nodeType"] == "NewExpression")
+		return createNewExpression(_json);
+	if (_json["nodeType"] == "MemberAccess")
+		return createMemberAccess(_json);
+	if (_json["nodeType"] == "IndexAccess")
+		return createIndexAccess(_json);
+	if (_json["nodeType"] == "Identifier")
+		return createIdentifier(createSourceLocation(_json), _json["nodeType"].asCString());
+	if (_json["nodeType"] == "ElementaryTypeNameExpression")
+		return createElementaryTypeNameExpression(_json);
+	if (_json["nodeType"] == "Literal")
+		return createLiteral(_json);
+	else BOOST_THROW_EXCEPTION(InternalCompilerError() << errinfo_comment("type of JsonValue is unknown."));
+}
+
+ASTPointer<SourceUnit> ASTJsonImporter::createSourceUnit(Json::Value const& _node)
+{
+	SourceLocation const& location = createSourceLocation(_node);
+	vector<ASTPointer<ASTNode>> nodes;
+	for (auto& child: _node["nodes"])
+		nodes.push_back(convertJsonToASTNode(child));
+	ASTPointer<SourceUnit> tmp = make_shared<SourceUnit>(location, nodes);
+	tmp->annotation().path = m_currentSource;
+	tmp->setID(_node["id"].asInt());
```

Comment:
Since `id` and `location` is part of every ast node, it might make sense to create a template function:
```
template <typename T, typename... Args>
ASTPointer<T> createASTNode(Json::Value const& _node, Args&&... _args)
{
    auto n = make_shared<T>(createSourceLocation(_node), forward<Args>(_args)...);
    n->setID(_node["id"].asInt());
    return n;
}
```

and then you would use it as:

```
ASTPointer<SourceUnit> ASTJsonImporter::createSourceUnit(Json::Value const& _node)
{
	vector<ASTPointer<ASTNode>> nodes;
	for (auto& child: _node["nodes"])
		nodes.emplace_back(convertJsonToASTNode(child));
	ASTPointer<SourceUnit> tmp = createASTNode<SourceUnit>(_node, nodes);
	tmp->annotation().path = m_currentSource;
        return tmp;
}
```